src/OrchardCore/OrchardCore/Shell/Removing/ShellRemovingManager.cs

```diff
@@ -0,0 +1,159 @@
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using OrchardCore.Environment.Shell.Builders;
+using OrchardCore.Environment.Shell.Models;
+using OrchardCore.Modules;
+
+namespace OrchardCore.Environment.Shell.Removing;
+
+public class ShellRemovingManager : IShellRemovingManager
+{
+    private readonly IShellHost _shellHost;
+    private readonly IShellContextFactory _shellContextFactory;
+    private readonly IEnumerable<IShellRemovingHandler> _shellRemovingHandlers;
+    private readonly ILogger _logger;
+
+    public ShellRemovingManager(
+        IShellHost shellHost,
+        IShellContextFactory shellContextFactory,
+        IEnumerable<IShellRemovingHandler> shellRemovingHandlers,
+        ILogger<ShellRemovingManager> logger)
+    {
+        _shellHost = shellHost;
+        _shellContextFactory = shellContextFactory;
+        _shellRemovingHandlers = shellRemovingHandlers;
+        _logger = logger;
+    }
+
+    public async Task<ShellRemovingContext> RemoveAsync(ShellSettings shellSettings, bool localResourcesOnly = false)
+    {
+        var context = new ShellRemovingContext
+        {
+            ShellSettings = shellSettings,
+            LocalResourcesOnly = localResourcesOnly,
+        };
+
+        if (shellSettings.Name == ShellHelper.DefaultShellName)
+        {
+            context.ErrorMessage = $"The tenant should not be the '{ShellHelper.DefaultShellName}' tenant.";
+            return context;
+        }
+
+        if (shellSettings.State != TenantState.Disabled && shellSettings.State != TenantState.Uninitialized)
+        {
+            context.ErrorMessage = $"The tenant '{shellSettings.Name}' should be 'Disabled' or 'Uninitialized'.";
+            return context;
+        }
+
+        // Check if the tenant is not 'Uninitialized' and that all resources should be removed.
+        if (shellSettings.State == TenantState.Disabled && !context.LocalResourcesOnly)
+        {
+            // Create an isolated shell context composed of all features that have been installed.
+            using var shellContext = await _shellContextFactory.CreateMaximumContextAsync(shellSettings);
+            (var locker, var locked) = await shellContext.TryAcquireShellRemovingLockAsync();
+            if (!locked)
+            {
+                _logger.LogError(
+                    "Failed to acquire a lock before executing the tenant handlers while removing the tenant '{TenantName}'.",
+                    shellSettings.Name);
+
+                context.ErrorMessage = $"Failed to acquire a lock before executing the tenant handlers.";
+                return context;
+            }
+
+            await using var acquiredLock = locker;
+
+            await shellContext.CreateScope().UsingServiceScopeAsync(async scope =>
+            {
+                // Execute tenant level removing handlers (singletons or scoped) in a reverse order.
```

Comment:
Why the reverse order? Add to the comment.

Reply:
Okay, will comment it, need to remember the singularity due to the removal process

Reply:
Okay done